### PR TITLE
Backward compatible usage of the new __serialize/__deserialize methods

### PR DESCRIPTION
**Pull request by goetas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Doc updated   | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT

### CODE REVIEW

**Comment by goetas** [comment]
@alexander-schranz  @W0rma  could you please try this?

**Comment by ruudk** [comment]
I also noticed this problem on PHP 8.1 and was waiting for a fix. I just tried out this branch, it did solve the problem. 

Withe the previous fix, it would only work when there was no cache, but as soon as it was read from cache it would break. Now it works as expected 🎉

**Comment by ruudk** [comment]
@goetas Thanks for merging this. It would be great to have this tagged so that I can [continue testing SimpleBus on PHP 8.1](https://github.com/SimpleBus/SimpleBus/pull/127) 🙏

**Comment by goetas** [comment]
https://github.com/schmittjoh/metadata/releases/tag/2.5.2

**Comment by wachterjohannes** [comment]
@goetas we have "tested" your new release (2.5.1 green 2.5.2 red) thru our github workflow (https://github.com/sulu/sulu/runs/4222298775?check_suite_focus=true) and it seems to have a bug which causes a segfault during the serialization process. the segfault happens inside of this function https://github.com/massiveart/MassiveSearchBundle/blob/2.x/Search/Metadata/ClassMetadata.php#L83 . do you have any idea what the problem could be?

**Comment by goetas** [comment]
hmm a segfault... that is scary :( , no idea :(

**Comment by alexander-schranz** [comment]
The segmentation fault, seems also happening here: https://github.com/massiveart/MassiveSearchBundle/runs/4265624083?check_suite_focus=true

**Comment by alexander-schranz** [comment]
@goetas it also very strange on 3v4l it ends in a memory limit: https://3v4l.org/CmKHo

File: src/PropertyMetadata.php [hunk 1]
```diff
@@ -53,25 +56,22 @@ public function serialize()
...
-     * @return array<string>
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingTraversableTypeHintSpecification
      */
     public function __serialize(): array
     {
-        return [
-            $this->class,
-            $this->name,
-        ];
+        return [$this->serialize()];
```
Comment (alexander-schranz) [inline src/PropertyMetadata.php:67]: this seems to trigger the parent `$this->serialize` method which will then end in a segmentation fault. Instead `$this->serialize` I think we need to go with implementing it seperatly. Still this would crash applicatons not implementing `__serialize` and `__unserialize` method so I would still suggest great a new major release and revert this changes.

**Comment by goetas** [comment]
@alexander-schranz  your custom methods are wrong:

```php
    public function __serialize(): array
    {
        return $this->serialize();
    }

    public function __unserialize(array $data): void
    {
        $this->unserialize($data);
    }
```

should be

```php

    public function __serialize()
    {
        $data = parent::__serialize();

        return [$data, \serialize($this->indexMetadatas), $this->repositoryMethod];
    }

    public function __unserialize(array $data)
    {
        list($data, $indexMetadata, $this->repositoryMethod) = $data;
        $this->indexMetadatas = \unserialize($indexMetadata);
        parent::__unserialize($data);
    }

    public function serialize(): array
    {
        return serialize($this->__serialize());
    }

    public function __unserialize(array $data): void
    {
        $this->unserialize($data[0]);
    }

```

see https://3v4l.org/ll2IK for a working version

**Comment by alexander-schranz** [comment]
@goetas still the same result it ends then in a segmentation fault: https://github.com/massiveart/MassiveSearchBundle/pull/162

File: src/PropertyMetadata.php (hunk 1 above)
Comment (goetas) [reply src/PropertyMetadata.php:67]: if we do a major release there is no way to allow applications to support both v2 and v3 versions of this library.

while with this implementation people can just add to their code

```
    public function __serialize(): array
    {
        return [$this->serialize()];
    }

    public function __unserialize(array $data): void
    {
        $this->unserialize($data[0]);
    }
```

and they got it working, or if they do not care about deprecations, it works anyway even without adding this code.
